hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/PrngSystemPrecompiledContract.java

```diff
@@ -0,0 +1,283 @@
+import static com.hedera.services.state.EntityCreator.EMPTY_MEMO;
+import static com.hedera.services.store.contracts.precompile.codec.EncodingFacade.SUCCESS_RESULT;
+import static com.hedera.services.store.contracts.precompile.codec.EncodingFacade.resultFrom;
+import static com.hedera.services.store.contracts.precompile.utils.PrecompilePricingUtils.GasCostType.PRNG;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.FAIL_INVALID;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INSUFFICIENT_GAS;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INVALID_PRNG_RANGE;
+
+/**
+ * System contract to generate random numbers. This will generate 256-bit pseudorandom number when no range is provided
+ * using n-3 record's running hash.
+ * If 32-bit integer "range" and 256-bit "seed" is provided returns a pseudorandom 32-bit integer X belonging to [0,
+ * range).
+ */
+@Singleton
+public class PrngSystemPrecompiledContract extends AbstractPrecompiledContract {
+	private static final Logger log = LogManager.getLogger(PrngSystemPrecompiledContract.class);
+	private static final String PRECOMPILE_NAME = "PRNG";
+	private static final LongType WORD_DECODER = TypeFactory.create("uint32");
+	//random256BitGenerator(uint256)
+	static final int PSEUDORANDOM_SEED_GENERATOR_SELECTOR = 0xd83bf9a1;
+	//randomNumberGeneratorInRange(uint32)
+	static final int PSEUDORANDOM_NUM_IN_RANGE_GENERATOR_SELECTOR = 0xb781b004;
+	public static final String PRNG_PRECOMPILE_ADDRESS = "0x169";
+	private final PrngLogic prngLogic;
+	private final SideEffectsTracker tracker;
+	private final EntityCreator creator;
+	private HederaStackedWorldStateUpdater updater;
+	private final RecordsHistorian recordsHistorian;
+	private final PrecompilePricingUtils pricingUtils;
+	private final Supplier<Instant> consensusNow;
+	private final LivePricesSource livePricesSource;
+	private final GlobalDynamicProperties properties;
+	private long gasRequirement;
+
+	@Inject
+	public PrngSystemPrecompiledContract(
+			final GasCalculator gasCalculator,
+			final PrngLogic prngLogic,
+			final ExpiringCreations creator,
+			final SideEffectsTracker tracker,
+			final RecordsHistorian recordsHistorian,
+			final PrecompilePricingUtils pricingUtils,
+			final Supplier<Instant> consensusNow,
+			final LivePricesSource livePricesSource,
+			final GlobalDynamicProperties properties) {
+		super(PRECOMPILE_NAME, gasCalculator);
+		this.prngLogic = prngLogic;
+		this.creator = creator;
+		this.tracker = tracker;
+		this.recordsHistorian = recordsHistorian;
+		this.pricingUtils = pricingUtils;
+		this.consensusNow = consensusNow;
+		this.livePricesSource = livePricesSource;
+		this.properties = properties;
+	}
+
+	@Override
+	public long gasRequirement(final Bytes bytes) {
+		return gasRequirement;
+	}
+
+	@Override
+	public PrecompileContractResult computePrecompile(final Bytes input, final MessageFrame frame) {
+		ExpirableTxnRecord.Builder childRecord;
```

Comment:
⚠️ We **must** prevent a `ContractCallLocal` query thread from entering this precompile; it will cause an ISS by corrupting the `RecordHistorian` singleton's state. The way to detect a `ContractCallLocal` thread is like [here](https://github.com/hashgraph/hedera-services/blob/da465efbfdafc19b5e6b29f29d8fbefcf65d2a5a/hedera-node/src/main/java/com/hedera/services/store/contracts/precompile/HTSPrecompiledContract.java#L173):
```
if (frame.isStatic()) {
    final var proxyUpdater = (HederaStackedWorldStateUpdater) frame.getWorldUpdater();
    if (!proxyUpdater.hasMutableLedgers()) {
        frame.setRevertReason(STATIC_CALL_REVERT_REASON);
        return Pair.of(defaultGas(), null);
    }
}
```

Reply:
We need to be able to have this functional as a static call.  We access block number, timestamp, and other values in static calls too.  Is there a way to read the information without corrupting the RecordHistorian?

Reply:
Fixed adding code for preventing a `ContractCallLocal` query thread from returning from this precompile. Also moved generation of randomNumber and returning Bytes to a separate function.